[OG-52] Spring security

LisieckaK00 opened the pull request:
- I moved `UserType` from `dto` package to `model` as it will now be used also for user roles.
- The `securityFilterChain` for now is really basic, and can't look like that in the future for sure. I let all requests to go to `/users` for now, to make it easy for us to add new users for tests. 

For requests that require authenticated users, you need to add `Authorization` header with token that is created by base64 encoding of string `<email>:<password>`. You may use for example: https://www.base64encode.org. Here is an examplary request:
```http
POST http://localhost:8080/rewards/chests
Content-Type: application/json
Authorization: Basic <token>

{
  "name": "chest1",
  "description": "description1",
  "imageUrl": "url1",
  "behavior": "ALL",
  "courseId": 1
}
```

LisieckaK00 commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/UserService.java:
```diff
@@ -7,17 +7,25 @@
 @AllArgsConstructor
-public class UserService {
+public class UserService implements UserDetailsService {
+    private final PasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
```
I'm not sure if this is a good idea, to have it in this way. However, I couldn't import the encoder as it created a loop in the dependencies. I also am not sure if we want to hash the password specifically in this service. Let me know if you have better ideas.

kligarski replied:
We can extract passwordEncoder to seperate configuration class, e.g. `AuthenticaitonConfig` or `CommonConfig`.

```java
@Configuration
public class AuthenticationConfig {
    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
}
```

and then use it autowire `passwordEncoder` for `SecurityConfig` and `UserService`. Maybe a little bit ugly but it's better than 2 sources of passwordEncoder.

We can probably also use `@Lazy`: https://www.baeldung.com/circular-dependencies-in-spring#2-use-lazy.

kligarski reviewed (changes requested):
When I started the application locally, the login does not work - I always get redirected to `/login` page even if I pass credentials in `Authorization` header. This is because we use `formLogin` instead of `httpBasic` - did you maybe add some setup to `application.properties` that makes it work on your side?

Either way, I think we need to change the current version. As we're using a React application, we cannot use login form provided by Spring Security. We shouldn't require passing credentials with every request as well - we don't want to store them in the browser for a long time.

I think we should use formLogin but disable default login page and provide `loginProcessingUrl`.

This code was generated but it is a nice **draft** of the idea (it obviously needs modifying endpoint URLs, probably responses as well):
```java
@Configuration
@EnableWebSecurity
public class SecurityConfig {

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http
            .csrf().disable() // or configure CSRF properly for APIs
            .cors().and()
            .authorizeHttpRequests(auth -> auth
                .requestMatchers("/api/login", "/api/logout", "/public/**").permitAll() // not sure if we need to permitAll /api/logout
                .anyRequest().authenticated()
            )
            .formLogin(form -> form
                .loginProcessingUrl("/api/login")  // Only handles POST /api/login
                .successHandler((request, response, authentication) -> {
                    response.setStatus(HttpServletResponse.SC_OK);
                })
                .failureHandler((request, response, exception) -> {
                    response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
                    response.setContentType("application/json");
                    response.getWriter().write("{\"error\": \"Invalid credentials\"}");
                })
                .permitAll() // not sure about this line, we need to check if this is correct
            )
            .exceptionHandling(ex -> ex
                .authenticationEntryPoint((request, response, authException) -> {
                    response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
                    response.setContentType("application/json");
                    response.getWriter().write("{\"error\": \"Unauthorized\"}");
                })
            )
            .logout(logout -> logout
                .logoutUrl("/api/logout")
                .logoutSuccessHandler((request, response, authentication) -> {
                    response.setStatus(HttpServletResponse.SC_OK);
                })
            );

        return http.build();
    }
}
```

Then we can pass credentals like this:
```
POST http://localhost:8080/api/login
Content-Type: application/x-www-form-urlencoded

username=a@b.com&password=password
```

Spring will automatically create a session - you can look into how it works and if we should tweak any settings ourselves.

We also **must implement CSRF proteciton and CORS**, but we might delay this until frontend-backend integration (but we should keep the changes seperate from develop branch - we will need CORS configuration for the frontend to be able to send requests).

I'm not a big fan of leaving `/users` available for everyone but maybe for a moment we can leave it like that.

LisieckaK00 commented:
1. When submitting the PR, I tested this the way I described earlier. Honestly, I’m not sure why I can’t reproduce it now — I must have made a mistake somewhere, though I’m not exactly sure what it was. Apologies for the confusion.

   After rechecking everything, I’m updating the comment to reflect how it actually works:

   First, you need to log in. As you mentioned, that should be done like this:
   ```
   POST http://localhost:8080/login
   Content-Type: application/x-www-form-urlencoded   
   
   username=a@b.com&password=password
   ```
   In response, the server sets a `JSESSIONID` cookie, which is automatically stored in the browser.

   Then, for all subsequent requests from the frontend, it's enough to add `credentials: "include"` to the fetch call. This ensures the session cookie is included in the request, and authentication works automatically — there’s no need to manually store credentials. For example:
   ```js
   await fetch("http://localhost:8080/rewards/chests/1/items", {   
     method: "GET",
     credentials: "include", // this sends the session cookie
   });
   ```
   If you’re testing via Postman or similar tools, you first need to log in via `/login`, copy the `JSESSIONID` from the response, and then include it in subsequent requests, like:
   ```   
   GET http://localhost:8080/rewards/chests/1/items
   Cookie: JSESSIONID=B750005F5E56E6B074A9BE91F97A5542
   ```
2. You're right about the login form — I'll make the necessary update.
3. Thinking more about the `/users` endpoints, I agree that it doesn't make much sense to leave them publicly accessible at this stage. I'll remove that in my next update.

LisieckaK00 commented:
I removed the `permitAll` for `/users` endpoint. However, it makes it difficult to test and run the app, so I added this again. Because of the password encoding, it's not easy to add users directly to the database. I don't have a better idea, so please let me know if you have any. If you all think that this shouldn't be available for everyone, I'll remove it again.

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/config/CorsConfig.java:
```diff
@@ -0,0 +1,30 @@
+    @Bean
+    public CorsConfigurationSource corsConfigurationSource() {
+        CorsConfiguration configuration = new CorsConfiguration();
+        configuration.setAllowedOriginPatterns(List.of(
+                "https://polymorphia-self.vercel.app",
+                "https://polymorphia-git-*.vercel.app",
+                "http://localhost:3000")
+        );
```
I'm thinking if we should leave it hardcoded in code or maybe extract it to `application.properties`?

Here is the example from one of the projects I did:
- `application.properties` (in out repo we have `application.properties.example`) [here](https://github.com/Trechu/cinemanager/blob/master/cinemanager/src/main/resources/application.properties)
- defining conf properties [here](https://github.com/Trechu/cinemanager/blob/master/cinemanager/src/main/java/pl/edu/agh/to/cinemanager/configuration/CorsProperties.java)
- using in the config [here](https://github.com/Trechu/cinemanager/blob/master/cinemanager/src/main/java/pl/edu/agh/to/cinemanager/configuration/SecurityConfig.java#L119)
- annotation in main app file [here](https://github.com/Trechu/cinemanager/blob/master/cinemanager/src/main/java/pl/edu/agh/to/cinemanager/CinemanagerApplication.java#L11)

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/config/SecurityConfig.java:
```diff
@@ -2,47 +2,83 @@
                 .authorizeHttpRequests(authorizeRequests -> authorizeRequests
-                        .requestMatchers("/users", "/users/*").permitAll()
+                        .requestMatchers(HttpMethod.POST, "/users").permitAll()
+                        .requestMatchers(HttpMethod.POST, "/courses/**", "/rewards/**", "/users/**").hasAuthority(UserType.COORDINATOR.getAuthority())
```
In the future, we might consider using [per-method authorization](https://docs.spring.io/spring-security/reference/servlet/authorization/method-security.html#use-preauthorize) if we get to more complex authorization scenarios but it is okay for now.

LisieckaK00 replied:
I totally agree, right now we don't have many endpoints, so I did that to limit the access a little. We may be updating this configuration when adding new endpoints that will need different type of access.

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/config/SecurityConfig.java:
```diff
@@ -2,47 +2,83 @@
         return httpSecurity
-                .cors(Customizer.withDefaults())
+                .cors(cors -> cors.configurationSource(corsConfig.corsConfigurationSource()))
                 .csrf(AbstractHttpConfigurer::disable)
```
I understand that we're leaving out CSRF for another PR?

LisieckaK00 replied:
Thanks for the cors, I must have missed that. I also forgot about the `csrf`, I'll try to do this in this PR.

kligarski reviewed (commented):
Great job! I have some suggestions for the future but I'm happy to leave them for later if merging the changes now would help with the development process.

As for creating users, maybe it's possible to use site like https://bcrypt-generator.com/ or create a simple python script/java console app to get the hashed password and then it would be possible to insert it directly into the database?

krudny commented on polymorphia-frontend/components/form/schema.ts:
```diff
@@ -0,0 +1,11 @@
+import { z } from "zod";
+
+export const emailField = z.string().email("Nieprawidłowy email");
+export const passwordField = z
+  .string()
+  .min(8, "Hasło musi mieć przynajmniej 8 znaków.");
```
It is not trivial, but possible to do.

LisieckaK00 replied:
It is actually quite easy with tanstack form. I updated it and now an error occurs after unfocusing a field.

krudny commented on polymorphia-frontend/services/AuthService.ts:
```diff
@@ -0,0 +1,23 @@
+import API_HOST from "./api";
+
+const AuthService = {
+  login: async (email: string, password: string) => {
```
Please, use JSON, not URLParams.

```js
headers: new Headers({ "Content-Type": "application/json",}),
body: JSON.stringify({
  username: email,
  password: password,
}),
```

LisieckaK00 replied:
Currently, we're using a default `/login` endpoint provided by spring security, that expects `"Content-Type": "application/x-www-form-urlencoded"`.  So you want me to create my own implementation of login that will accept json, right?

kligarski replied:
It also caught me off guard but with `"Content-Type": "application/x-www-form-urlencoded"` you pass the data in the same format as `URLSearchParams` but they are not actually GET search params. I think we're good but let's wait for @krudny's answer.

krudny replied:
I'm not entirely sure why Spring Security use by default the `application/x-www-form-urlencoded` format for the `/login` endpoint; it does seem a bit outdated compared to accepting JSON in the request body, which is common practice nowadays.

We could implement a custom authentication controller and service layer to handle JSON submissions, but I am not skilled enough to make deeper discussions about that.  Moreover, this isn't a critical change for now.

LisieckaK00 replied:
As we discussed, this will not be addressed in this PR.

krudny commented:
In #8 Krzysztof have changed a way we handle API_HOST variable, please include his idea in your work.

krudny commented on polymorphia-frontend/components/form/FieldInfo.tsx:
```diff
@@ -0,0 +1,15 @@
+import { AnyFieldApi } from "@tanstack/react-form";
```
I believe our app will contain many forms, so naming the folder 'form' might not be the most descriptive choice. Since it's closely related to /home and LoginForm.tsx, adding a separate folder feels like unnecessary overhead.

LisieckaK00 replied:
My idea was that `FieldInfo` and `schema.` are extracted to `form` folder, so that they can be reused in next forms that we create. `FieldInfo` is responsible for showing errors for a given field, which may be useful to be the same for different forms. When it comes to `schema`: I thought it may be similar to current `DTO` file and contain schemas for different forms. We could be reusing some of the fields, like `passwordField` for registration or changing the password. If you don't agree with my way of thinking, let me know, I'll move these files.

krudny replied:
Okey, leave them there.

LisieckaK00 commented on polymorphia-frontend/components/home/LoginForm.tsx:
```diff
@@ -1,61 +1,109 @@
+                <ButtonWithBorder
+                  text="Zaloguj się"
+                  className="mt-12"
+                  isActive={isPristine || !canSubmit}
```
You're right, it is quite confusing right now

LisieckaK00 replied:
As we discussed, this will not be addressed in this PR.

LisieckaK00 commented on polymorphia-frontend/components/home/LoginForm.tsx:
```diff
@@ -1,61 +1,109 @@
-    console.log('Hasło:', password);
-  };
+  const loginMutation = useMutation({
+    mutationFn: ({ email, password }: { email: string; password: string }) =>
+      AuthService.login(email, password),
```
I was already using in the `LoginForm` an interface `LoginDto` that is located with other dtos. I updated this logic to also use this interface. If you think it should have a different name or a different location or anything else, let me know, I'll update it.

ustankie commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/config/SecurityConfig.java:
```diff
@@ -0,0 +1,99 @@
+                .cors(cors -> cors.configurationSource(corsConfigurationSource()))
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(authorizeRequests -> authorizeRequests
+                        .requestMatchers(HttpMethod.POST, "/courses/**", "/rewards/**", "/users/**").hasAuthority(UserType.COORDINATOR.getAuthority())
```
I think we cannot categorize it this way, i.e. /courses/{id}/chests is needed for the "Baza wiedzy" page, so everyone should be able to access it, not only coordinator. I've seen the comment that @kligarski left and probably doing per method authorization would be the best option here

LisieckaK00 replied:
Here, only coordinator can execute POST methods. As this are things that he will be adding when configuring the course. Every authenticated user can access GET methods, so no problem with students and "Baza wiedzy"

LisieckaK00 replied:
Or did it work for you with user and post method?

LisieckaK00 replied:
As we discussed, works properly.
